mscore/timeline.cpp

```diff
@@ -0,0 +1,1522 @@
+                  }
+
+            _score->update();
+            mscore->endCmd();
+            }
+      else if (state == ViewState::DRAG) {
+            this->setCursor(Qt::ArrowCursor);
+            mscore->endCmd();
+            }
+      state = ViewState::NORMAL;
+      }
+
+//---------------------------------------------------------
+//   updateGrid
+//---------------------------------------------------------
+
+void Timeline::updateGrid()
+      {
+      if (_score) {
+            drawGrid(nstaves(), _score->nmeasures());
+            updateView(0, 0);
+            drawSelection();
+            }
+
+      viewport()->update();
+      }
+
+//---------------------------------------------------------
+//   setScore
+//---------------------------------------------------------
+
+void Timeline::setScore(Score* s)
+      {
+      _score = s;
+      scene()->clear();
+
+      if (_score) {
+            drawGrid(nstaves(), _score->nmeasures());
+            changeSelection(SelState::NONE);
+            }
+
+      viewport()->update();
+      }
+
+//---------------------------------------------------------
+//   setScoreView
+//---------------------------------------------------------
+
+void Timeline::setScoreView(ScoreView* v)
+      {
+      _cv = v;
+      if (_cv) {
+            connect(_cv, SIGNAL(offsetChanged(double, double)), this, SLOT(updateView(double, double)));
+            updateView(0, 0);
+            }
+      }
+
+//---------------------------------------------------------
+//   updateView
+//---------------------------------------------------------
+
+void Timeline::updateView(double, double)
+      {
+      if (_cv) {
+
+            std::set<std::pair<Measure*, int>> visible_items;
+            for (Measure* m = _score->firstMeasure(); m; m = m->nextMeasure()) {
+                  System* sys = m->system();
+                  for (int staff = 0; staff < _score->staves().length(); staff++) {
+                        if (!_score->staff(staff)->show())
+                              continue;
+                        QRectF stave = QRectF(sys->canvasBoundingRect().left(),
+                                                          sys->staffCanvasYpage(staff),
+                                                          sys->width(),
+                                                          sys->staff(0)->bbox().height());
+                        QRectF showRect = m->canvasBoundingRect().intersected(stave);
+                        QRectF canvas = QRectF(_cv->matrix().inverted().mapRect(_cv->geometry()));
+                        if (canvas.intersects(showRect)) {
+                              std::pair<Measure*, int> p(m, staff);
+                              visible_items.insert(p);
+                              }
+
+
+                        }
+                  }
+            QPainterPath pp = QPainterPath();
+            pp.setFillRule(Qt::WindingFill);
+            for (QGraphicsItem* i: scene()->items()) {
```

Comment:
I made the following comment in another place on github, but basically I would prefer "gi" here over "i" because "i" is a bit overdone and to me is hard to read.